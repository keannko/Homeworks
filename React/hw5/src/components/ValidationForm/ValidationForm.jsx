import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Formik, Form, Field } from "formik";
import * as Yup from "yup";
import s from "./ValidationForm.module.scss";
import { PatternFormat } from "react-number-format";

const DisplayingErrorMessagesSchema = Yup.object().shape({
  name: Yup.string().required("Required"),
  surname: Yup.string().required("Required"),
  age: Yup.number().typeError('Invalid Age').required("Required"),
  address: Yup.string().required("Required"),
  phone: Yup.string().required("Required"),
});

function validatePhone(value) {
  let error;
  if (value[13] === "#") {
    error = "More numbers!";
  }
  return error;
}

const ValidationForm = () => {
  const dispatch = useDispatch();
  const cartItems = useSelector((state) => state.cart.cart);
  const products = useSelector((state) => state.products.products);
  const order = useSelector((state) => state.cart.qty);

  const removeFromCart = () => {
    return function (dispatch) {
      cartItems.forEach((element) => {
        dispatch({ type: "REMOVE_FROM_CART", payload: { id: element } });
      });
    };
  };

  const createOrder = (prods, ord) => {
    const productsOrder = [];
    prods.forEach((el) => {
      if (Object.keys(ord).includes(el.id)) {
        const itemOrder = {
          product: el.name,
          price: el.price,
          qty: ord[el.id],
          sum: ord[el.id] * el.price,
        };
        productsOrder.push(itemOrder);
      }
    });
    console.log(productsOrder);
  };

  return (
    <div className={s.wrap}>
      <Formik
        initialValues={{
          name: "",
          surname: "",
          age: "",
          address: "",
          phone: "",
        }}
        validationSchema={DisplayingErrorMessagesSchema}
        onSubmit={(values) => {
          console.log(values);
          createOrder(products, order);
          dispatch(removeFromCart());
          alert(
            "Thank you for your order, check your email for order details!"
          );
        }}
      >
        {({ errors, touched }) => (
          <Form className={s.mainForm}>
            <div className={s.mainFormName}>
              <h4>Name: </h4>
              <Field name="name" />
              {touched.name && errors.name && <div>{errors.name}</div>}
            </div>
            <div className={s.mainFormSurname}>
              <h4>Surname: </h4>
              <Field name="surname" />
              {touched.surname && errors.surname && <div>{errors.surname}</div>}
            </div>
            <div className={s.mainFormAge}>
              <h4>Age: </h4>
              <Field name="age" />
              {touched.age && errors.age && <div>{errors.age}</div>}
            </div>
            <div className={s.mainFormAddress}>
              <h4>Address: </h4>
              <Field name="address" />
              {touched.address && errors.address && <div>{errors.address}</div>}
            </div>
            <div className={s.mainFormName}>
              <h4>Phone: </h4>
              <Field name="phone" validate={validatePhone}>
                {({ field }) => (
                  <div>
                    <PatternFormat
                      {...field}
                      format="(###)###-##-##"
                      allowEmptyFormatting
                      mask="#"
                    />
                    {errors.phone && touched.phone && (
                      <div className={s.errorMessage}>{errors.phone}</div>
                    )}
                  </div>
                )}
              </Field>
            </div>
            <button className={s.buttonForm} type="submit">
              Checkout
            </button>
          </Form>
        )}
      </Formik>
    </div>
  );
};

export default ValidationForm;
