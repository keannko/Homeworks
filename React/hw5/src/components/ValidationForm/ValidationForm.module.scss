.mainForm {
    display: grid;
    grid-template-columns: 1fr;
    gap: 20px;
}

.buttonForm {
    padding: 20px 10px;
    cursor: pointer;
    border-radius: 10px;
    background-color: lightgreen;
    font-size: 18px;
    font-weight: 600;
}