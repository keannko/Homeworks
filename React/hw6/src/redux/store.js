import { createStore, combineReducers, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { favoritesReducer } from "./reducers/favoritesReducer";
import { productsReducer } from "./reducers/productsReducer";
import { modalWindowReducer } from "./reducers/modalWindowReducer";
import { cartReducer } from "./reducers/cartReducer";

const rootReducer = combineReducers({
  favorites: favoritesReducer,
  products: productsReducer,
  modal: modalWindowReducer,
  cart: cartReducer
});

export const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(thunk))
);
