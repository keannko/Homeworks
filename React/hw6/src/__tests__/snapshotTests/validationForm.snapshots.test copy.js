import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "../../redux/store";
import ValidationForm from "../../components/ValidationForm/ValidationForm"

describe("ValidationForm component snapshot", () => {
  test("snapshot", () => {
    const renderValidationForm = render(
      <Provider store={store}>
        <BrowserRouter>
          <ValidationForm />
        </BrowserRouter>
      </Provider>
    );
    expect(renderValidationForm).toMatchSnapshot();
  });
});
