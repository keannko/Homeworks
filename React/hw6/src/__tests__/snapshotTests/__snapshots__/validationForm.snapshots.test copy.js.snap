// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationForm component snapshot snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="wrap"
      >
        <form
          action="#"
          class="mainForm"
        >
          <div
            class="mainFormName"
          >
            <h4>
              Name: 
            </h4>
            <input
              name="name"
              value=""
            />
          </div>
          <div
            class="mainFormSurname"
          >
            <h4>
              Surname: 
            </h4>
            <input
              name="surname"
              value=""
            />
          </div>
          <div
            class="mainFormAge"
          >
            <h4>
              Age: 
            </h4>
            <input
              name="age"
              value=""
            />
          </div>
          <div
            class="mainFormAddress"
          >
            <h4>
              Address: 
            </h4>
            <input
              name="address"
              value=""
            />
          </div>
          <div
            class="mainFormName"
          >
            <h4>
              Phone: 
            </h4>
            <div>
              <input
                inputmode="numeric"
                name="phone"
                type="text"
                value="(###)###-##-##"
              />
            </div>
          </div>
          <button
            class="buttonForm"
            type="submit"
          >
            Checkout
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="wrap"
    >
      <form
        action="#"
        class="mainForm"
      >
        <div
          class="mainFormName"
        >
          <h4>
            Name: 
          </h4>
          <input
            name="name"
            value=""
          />
        </div>
        <div
          class="mainFormSurname"
        >
          <h4>
            Surname: 
          </h4>
          <input
            name="surname"
            value=""
          />
        </div>
        <div
          class="mainFormAge"
        >
          <h4>
            Age: 
          </h4>
          <input
            name="age"
            value=""
          />
        </div>
        <div
          class="mainFormAddress"
        >
          <h4>
            Address: 
          </h4>
          <input
            name="address"
            value=""
          />
        </div>
        <div
          class="mainFormName"
        >
          <h4>
            Phone: 
          </h4>
          <div>
            <input
              inputmode="numeric"
              name="phone"
              type="text"
              value="(###)###-##-##"
            />
          </div>
        </div>
        <button
          class="buttonForm"
          type="submit"
        >
          Checkout
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
